src/dataflow/src/render/top_k.rs

```diff
@@ -40,38 +40,43 @@ where
-                // monotonic with an ENVELOPE of NONE, which is the default for ENVELOPE in
-                // Materialize and commonly used by users.
-                let result = self.render_top1_monotonic(ok_input, group_key, order_key);
-                self.collections
-                    .insert(relation_expr.clone(), (result, err_input));
-            } else if *monotonic {
-                // For monotonic inputs, we are able to retract inputs not produced as outputs.
-                use differential_dataflow::operators::iterate::Variable;
-                let delay = std::time::Duration::from_nanos(10_000_000_000);
-                let retractions = Variable::new(&mut ok_input.scope(), delay.as_millis() as u64);
-                let thinned = ok_input.concat(&retractions.negate());
-                let result = build_topk(thinned, group_key, order_key, *offset, *limit, arity);
-                retractions.set(&ok_input.concat(&result.negate()));
-                self.collections
-                    .insert(relation_expr.clone(), (result, err_input));
-            } else {
-                let result = build_topk(ok_input, group_key, order_key, *offset, *limit, arity);
-                self.collections
-                    .insert(relation_expr.clone(), (result, err_input));
-            }
+            // We create a new region to compartmentalize the topk logic.
+            let ok_result = ok_input.scope().region_named("TopK", |inner| {
+                let ok_input = ok_input.enter(inner);
+                let ok_result = if *monotonic && *offset == 0 && *limit == Some(1) {
+                    // If the input to a TopK is monotonic (aka append-only aka no retractions) then we
+                    // don't have to worry about keeping every row we've seen around forever. Instead,
+                    // the reduce can incrementally compute a new answer by looking at just the old TopK
+                    // for a key and the incremental data.
+                    //
+                    // This optimization generalizes to any TopK over a monotonic source, but we special
+                    // case only TopK with offset=0 and limit=1 (aka Top1) for now. This is because (1)
+                    // Top1 can merge in each incremental row in constant space and time while the
+                    // generalized solution needs something like a priority queue and (2) we expect Top1
+                    // will be a common pattern used to turn a Kafka source's "upsert" semantics into
+                    // differential's semantics. (2) is especially interesting because Kafka is
+                    // monotonic with an ENVELOPE of NONE, which is the default for ENVELOPE in
+                    // Materialize and commonly used by users.
+                    render_top1_monotonic(ok_input, group_key, order_key)
+                } else if *monotonic && *offset == 0 {
```

Comment:
Comment explaining what's going on with the `*offset == 0` plz

Reply:
Added, and a TODO about how it could be relaxed.